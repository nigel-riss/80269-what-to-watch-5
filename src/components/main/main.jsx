import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {
  getGenreList,
} from '../../utils/common.js';
import {
  incrementItemsShownCount,
  resetItemsShownCount,
  setGenre,
} from '../../store/actions/actions.js';
import {selectFilmsByGenre} from '../../store/selector.js';
import filmType from '../../types/film.js';
import FilmList from '../film-list/film-list.jsx';
import Footer from '../footer/footer.jsx';
import GenreList from '../genre-list/genre-list.jsx';
import MoreButton from '../more-button/more-button.jsx';
import PromoFilm from '../promo-film/promo-film';


const Main = (props) => {
  const {
    activeGenre,
    films,
    filmsByGenre,
    itemsShownCount,
    onGenreSelect,
    onMoreButtonClick,
    promoFilm,
  } = props;

  return (
    <React.Fragment>
      {promoFilm && <PromoFilm
        film={promoFilm}
      />}

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          <GenreList
            activeGenre={activeGenre}
            genres={getGenreList(films)}
            onGenreItemClick={(genre) => {
              onGenreSelect(genre);
            }}
          />

          <FilmList
            films={filmsByGenre.slice(0, itemsShownCount)}
          />

          {(filmsByGenre.length > itemsShownCount) && <MoreButton
            onClick={onMoreButtonClick}
          />}
        </section>

        <Footer/>
      </div>
    </React.Fragment>
  );
};


Main.propTypes = {
  activeGenre: PropTypes.string,
  films: PropTypes.arrayOf(filmType),
  filmsByGenre: PropTypes.arrayOf(filmType),
  itemsShownCount: PropTypes.number.isRequired,
  onGenreSelect: PropTypes.func.isRequired,
  onMoreButtonClick: PropTypes.func.isRequired,
  promoFilm: filmType,
};


const mapStateToProps = ({APP, DATA}) => ({
  activeGenre: APP.genre,
  films: DATA.films,
  filmsByGenre: selectFilmsByGenre({APP, DATA}),
  promoFilm: DATA.promoFilm,
  itemsShownCount: APP.itemsShownCount,
});


const mapDispatchToProps = (dispatch) => ({
  onGenreSelect(genre) {
    dispatch(setGenre(genre));
    dispatch(resetItemsShownCount());
  },

  onMoreButtonClick() {
    dispatch(incrementItemsShownCount());
  },
});


export {Main};
export default connect(mapStateToProps, mapDispatchToProps)(Main);
