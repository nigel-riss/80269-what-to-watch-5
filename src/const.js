const FilmTab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

const RatingLimits = {
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  VERY_GOOD: 10,
};

const HOVER_PLAY_DELAY = 1000;
const FILM_CARD_HEIGHT = 175;
const FILM_CARD_WIDTH = 280;
const FILMS_SHOWN_PER_CLICK = 8;

export {
  FILM_CARD_HEIGHT,
  FILM_CARD_WIDTH,
  FILMS_SHOWN_PER_CLICK,
  FilmTab,
  HOVER_PLAY_DELAY,
  RatingLimits,
};
