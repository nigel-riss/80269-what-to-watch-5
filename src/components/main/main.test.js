import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import {Main} from './main.jsx';
import filmsMock from '../../mocks/films.js';
import storeMock from '../../mocks/store';
import {FILMS_SHOWN_PER_CLICK} from '../../const.js';


const mockStore = configureMockStore([]);
const store = mockStore(storeMock);
const noop = () => {};

it(`Should Main render correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <Main
              activeGenre={`Action`}
              films={filmsMock}
              filmsByGenre={filmsMock.slice(0, 4)}
              itemsShownCount={FILMS_SHOWN_PER_CLICK}
              onGenreSelect={noop}
              onMoreButtonClick={noop}
              promoFilm={filmsMock[0]}
            />
          </MemoryRouter>
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        }
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
