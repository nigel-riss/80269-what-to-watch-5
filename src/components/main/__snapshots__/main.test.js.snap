// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Main render correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Aviator"
        src="/img/aviator.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Aviator"
            height="327"
            src="/img/aviator-poster.png"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Aviator
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Biography
            </span>
            <span
              className="movie-card__year"
            >
              2004
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="14"
                viewBox="0 0 18 14"
                width="18"
              >
                <use
                  xlinkHref="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Adventure
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Biography
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
            onClick={[Function]}
          >
            Comedy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#ffffff",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <video
              disabled={true}
              height={175}
              muted={true}
              poster="/url/to/image.png"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              The Aviator
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#000000",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <video
              disabled={true}
              height={175}
              muted={true}
              poster="/url/to/image.png"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#ff00ff",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <video
              disabled={true}
              height={175}
              muted={true}
              poster="/url/to/image.png"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              Bohemian Rhapsody
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "backgroundColor": "#ff0000",
            }
          }
        >
          <div
            className="small-movie-card__image"
          >
            <video
              disabled={true}
              height={175}
              muted={true}
              poster="/url/to/image.png"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
            >
              Fantastic Beasts: The Crimes of Grindelwald
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
